File: 993. Cousins in Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    Set <Integer> parents;
    Set <Integer> depths;
    
    public boolean isCousins(TreeNode root, int x, int y) {
        parents = new HashSet();
        depths = new HashSet();
        traverseTree(root, x, y, root.val, 0);
        if(parents.size() != 1 && depths.size() == 1){
            return true;
        }
        return false;
    }
    
    private void traverseTree(TreeNode tree, int x, int y, int parent, int depth) {
        if(tree == null){
            return;
        }
        if(x == tree.val || y == tree.val){
            parents.add(parent);
            depths.add(depth);
        }
        traverseTree(tree.right, x, y, tree.val, depth+1);
        traverseTree(tree.left, x, y, tree.val, depth+1);
    }
}

File: javascript/993. Cousins in Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} x
 * @param {number} y
 * @return {boolean}
 */
var isCousins = function(root, x, y) {
    //deapth of x and y should be the same
    //their parents should not be the same
    let parents = [];
    let depths = [];
    function traverseTree(tree, depth, parent){
        if(tree == null){
            return;
        }
        if(tree.val == x || tree.val == y){
            parents.push(parent);
            depths.push(depth);
        }
        return traverseTree(tree.right, depth+1, tree.val) || traverseTree(tree.left, depth+1, tree.val);
        
    }
    traverseTree(root, 0, root.val);
    return (parents[0] != parents[1]) && (depths[0] == depths[1]);
};
